crates/adapters/src/format/parquet/mod.rs

```diff
@@ -0,0 +1,367 @@
+use parquet::file::reader::{FileReader, SerializedFileReader};
+use serde::Deserialize;
+use serde_arrow::schema::SerdeArrowSchema;
+use serde_arrow::ArrowBuilder;
+use serde_urlencoded::Deserializer as UrlDeserializer;
+use serde_yaml::Value as YamlValue;
+
+use crate::catalog::CursorWithPolarity;
+use crate::format::MAX_DUPLICATES;
+use crate::{
+    catalog::{DeCollectionStream, InputCollectionHandle, RecordFormat, SerBatch},
+    format::{Encoder, InputFormat, OutputFormat, ParseError, Parser},
+    ControllerError, OutputConsumer, SerCursor,
+};
+use pipeline_types::format::json::JsonFlavor;
+use pipeline_types::format::parquet::{ParquetEncoderConfig, ParquetParserConfig};
+use pipeline_types::program_schema::{ColumnType, Relation};
+
+#[cfg(test)]
+mod test;
+
+/// CSV format parser.
+pub struct ParquetInputFormat;
+
+impl InputFormat for ParquetInputFormat {
+    fn name(&self) -> Cow<'static, str> {
+        Cow::Borrowed("parquet")
+    }
+
+    /// Create a parser using configuration extracted from an HTTP request.
+    // We could just rely on serde to deserialize the config from the
+    // HTTP query, but a specialized method gives us more flexibility.
+    fn config_from_http_request(
+        &self,
+        _endpoint_name: &str,
+        _request: &HttpRequest,
+    ) -> Result<Box<dyn ErasedSerialize>, ControllerError> {
+        Ok(Box::new(ParquetParserConfig {}))
+    }
+
+    fn new_parser(
+        &self,
+        _endpoint_name: &str,
+        input_stream: &InputCollectionHandle,
+        _config: &YamlValue,
+    ) -> Result<Box<dyn Parser>, ControllerError> {
+        let input_stream = input_stream
+            .handle
+            .configure_deserializer(RecordFormat::Json(JsonFlavor::Default))?;
+        Ok(Box::new(ParquetParser::new(input_stream)) as Box<dyn Parser>)
+    }
+}
+
+struct ParquetParser {
+    /// Input handle to push parsed data to.
+    input_stream: Box<dyn DeCollectionStream>,
+    buf: Vec<u8>,
+    last_event_number: u64,
+}
+
+impl ParquetParser {
+    fn new(input_stream: Box<dyn DeCollectionStream>) -> Self {
+        Self {
+            input_stream,
+            buf: Vec::with_capacity(4096),
+            last_event_number: 0,
+        }
+    }
+}
+
+impl Parser for ParquetParser {
+    fn input_fragment(&mut self, data: &[u8]) -> (usize, Vec<ParseError>) {
+        self.buf.extend_from_slice(data);
+        (0, vec![])
+    }
+
+    fn eoi(&mut self) -> (usize, Vec<ParseError>) {
+        let bytes = Bytes::from(take(&mut self.buf));
+        match SerializedFileReader::new(bytes) {
+            Ok(reader) => {
+                let (mut cnt, mut errors) = (0, vec![]);
+                match reader.get_row_iter(None) {
+                    Ok(iter) => {
+                        for maybe_record in iter {
+                            match maybe_record {
+                                Ok(record) => {
+                                    let record_json = record.to_json_value().to_string();
```

Comment:
This probably needs a comment explaining what's going on and a TODO to do it properly